wicketd/src/update_planner.rs

```diff
@@ -0,0 +1,489 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use crate::artifacts::WicketdArtifactStore;
+use crate::mgs::make_mgs_client;
+use crate::update_events::UpdateEventFailureKind;
+use crate::update_events::UpdateEventKind;
+use crate::update_events::UpdateEventSuccessKind;
+use crate::update_events::UpdateStateKind;
+use anyhow::bail;
+use anyhow::ensure;
+use anyhow::Context;
+use buf_list::BufList;
+use debug_ignore::DebugIgnore;
+use gateway_client::types::SpIdentifier;
+use gateway_client::types::SpType;
+use gateway_client::types::SpUpdateStatus;
+use gateway_client::types::UpdateBody;
+use gateway_messages::SpComponent;
+use omicron_common::api::internal::nexus::KnownArtifactKind;
+use omicron_common::update::ArtifactId;
+use slog::error;
+use slog::info;
+use slog::o;
+use slog::Logger;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::net::SocketAddrV6;
+use std::sync::Arc;
+use std::sync::Mutex as StdMutex;
+use std::time::Duration;
+use std::time::Instant;
+use thiserror::Error;
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use uuid::Uuid;
+
+// These three types are mirrors of the HTTP
+// `UpdateState`/`UpdateEvent`/`UpdateLog`, but with the timestamps stored as
+// `Instant`s. This allows us to convert them to `Duration`s (i.e., ages) when
+// returning them to a caller.
+#[derive(Clone, Debug)]
+struct UpdateState {
+    timestamp: Instant,
+    kind: UpdateStateKind,
+}
+
+impl From<UpdateState> for crate::update_events::UpdateState {
+    fn from(state: UpdateState) -> Self {
+        Self { age: state.timestamp.elapsed(), kind: state.kind }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct UpdateEvent {
+    timestamp: Instant,
+    kind: UpdateEventKind,
+}
+
+impl From<UpdateEvent> for crate::update_events::UpdateEvent {
+    fn from(event: UpdateEvent) -> Self {
+        Self { age: event.timestamp.elapsed(), kind: event.kind }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct UpdateLog {
+    current: Option<UpdateState>,
+    events: Vec<UpdateEvent>,
+}
+
+impl From<UpdateLog> for crate::update_events::UpdateLog {
+    fn from(log: UpdateLog) -> Self {
+        Self {
+            current: log.current.map(Into::into),
+            events: log.events.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct UpdatePlanner {
+    mgs_client: gateway_client::Client,
+    artifact_store: WicketdArtifactStore,
```

Comment:
Seems like the planner doesn't need to store a copy of the artifact store -- it can just receive it in `regenerate_plan`.